### PR TITLE
Feature/forward message to new topic

### PR DESCRIPTION
**Pull request by izzyblues**
As a new feature, now you can forward to a new topic the consumed message if the ingestion is successful.

This PR includes additional changes:
- General refactoring to improve code readability and unit testing.
- Usage of new consumer and producer libraries to reuse code for health check (see https://github.com/Financial-Times/message-queue-gonsumer/pull/29 and https://github.com/Financial-Times/message-queue-go-producer/pull/6)
- Introduction of Logrus for logging
- Introduction of Build-info endpoint
- Upgrade to alpine 3.5 and Go 1.7.3 for docker image
- Introduction of Coverall and codecov.io

### CODE REVIEW

File: queue/message_handler_test.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	mh := NewMessageHandler(w)
+	mh.ForwardTo(p)
+	mh.HandleMessage(goodMsg)
+
+	w.AssertExpectations(t)
+	p.AssertNotCalled(t, "SendMessage", mock.AnythingOfType("string"), mock.AnythingOfType("producer.Message"))
+}
+
+func TestForwardFailBecauseOfProducer(t *testing.T) {
+	hook := testLog.NewGlobal()
```
Comment (khatton-ft) [inline queue/message_handler_test.go]: :+1:

File: queue/publication_event.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	body := make(native.ContentBody)
+	if err := json.Unmarshal([]byte(pe.Body), &body); err != nil {
+		return native.ContentBody{}, err
+	}
+
+	timestamp, found := pe.Headers["Message-Timestamp"]
+	if !found {
+		return native.ContentBody{}, errors.New("Publish event does not contain timestamp")
+	}
+	body["lastModified"] = timestamp
```
Comment (khatton-ft) [inline queue/publication_event.go]: Do we need to do any validation on the value?

File: Dockerfile [hunk 3]
```diff
@@ -1,19 +1,26 @@
...
-  && apk --update add go \
+RUN apk --update add git go libc-dev \
   && export GOPATH=/gopath \
   && REPO_PATH="github.com/Financial-Times/native-ingester" \
   && mkdir -p $GOPATH/src/${REPO_PATH} \
   && cp -r native-ingester/* $GOPATH/src/${REPO_PATH} \
   && cd $GOPATH/src/${REPO_PATH} \
+  && BUILDINFO_PACKAGE="github.com/Financial-Times/service-status-go/buildinfo." \
+  && VERSION="version=$(git describe --tag --always 2> /dev/null)" \
+  && echo here \
```
Comment (izzyblues) [inline Dockerfile]: Ops... It is how I debug Dockerfiles... :P

File: README.md [hunk 4]
```diff
@@ -1,29 +1,56 @@
 Native Ingester
 ===============
+[![CircleCI](https://circleci.com/gh/Financial-Times/native-ingester.svg?style=svg)](https://circleci.com/gh/Financial-Times/native-ingester) [![Go Report Card](https://goreportcard.com/badge/github.com/Financial-Times/native-ingester)](https://goreportcard.com/report/github.com/Financial-Times/native-ingester) [![Coverage Status](https://coveralls.io/repos/github/Financial-Times/native-ingester/badge.svg?branch=master)](https://coveralls.io/github/Financial-Times/native-ingester?branch=master) [![codecov](https://codecov.io/gh/Financial-Times/native-ingester/branch/master/graph/badge.svg)](https://codecov.io/gh/Financial-Times/native-ingester)
 
-* Ingests native content (articles, metadata, etc.) and persists them in the native-store.
-* Reads from ONE queue topic, adds a timestamp to the message and writes to ONE db collection.
+Native ingester implements the following functionality:
+1. It consumes messages containing native CMS content or native CMS metadata from ONE queue topic.
+1. According to the data source, native ingester writes the content or metadata in a specific db collection.
```
Comment (izzyblues) [inline README.md]: Sorry, this is "Italinglish".

File: app.go [hunk 5]
```diff
@@ -10,112 +10,147 @@ import (
...
+	writeQueueTopic := app.String(cli.StringOpt{
+		Name:   "write-topic",
+		Value:  "",
+		Desc:   "The topic to write the meassages to.",
+		EnvVar: "Q_WRITE_TOPIC",
+	})
+	writeQueueHostHeader := app.String(cli.StringOpt{
+		Name:   "write-queue-host-header",
+		Value:  "kafka",
+		Desc:   "The host header for the queue to write the meassages to.",
```
Comment (izzyblues) [inline app.go]: Ummmm... I think I should fix an appointment with my analyst.

File: native/content_body_parser.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+func NewContentBodyParser(uuidJSONPaths []string) ContentBodyParser {
+	return &contentBodyParser{uuidJSONPaths}
+}
+
+func (p contentBodyParser) getUUID(body map[string]interface{}) (string, error) {
+	jq := jsonq.NewQuery(body)
+	for _, uuidPath := range p.uuidJSONPaths {
+		uuid, err := jq.String(strings.Split(uuidPath, ".")...)
+		if err == nil && uuid != "" {
+			return uuid, nil
```
Comment (izzyblues) [inline native/content_body_parser.go]: Sounds good to me

File: native/native_writer.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+	response, err := nw.httpClient.Do(request)
+
+	if err != nil {
+		log.WithError(err).WithField("transaction_id", cBody.publishReference()).WithField("requestURL", requestURL).Error("Error calling native writer. Ignoring message.")
+		return err
+	}
+	defer properClose(response)
+
+	if response.StatusCode != http.StatusOK {
```
Comment (izzyblues) [inline native/native_writer.go]: Okay!

File: queue/message_handler_test.go (hunk 1 above)
Comment (izzyblues) [inline queue/message_handler_test.go]: :metal:

File: queue/publication_event.go (hunk 2 above)
Comment (izzyblues) [inline queue/publication_event.go]: In theory, It is something generated by cms-notifier, but if you want to make it bulletproof I can implement such check as well.

**Comment by izzyblues** [comment]
@khatton-ft I decided to put the source code in the vendor folder because it is an "agnostic" way for vendoring. 
Currently, there is not a shared agreement on vendoring tools for Go by the developers community.
In addition, the Go Team is working on a [official solution](https://github.com/golang/dep) to the vendoring problem.

I know, that adding the source code of a project is very ugly, but does it make sense to create a depency in Dockefiles and CirlceCi to a vendoring tool that probably is going to be deprecated in the next future?
(By the way, I used govendor to generate the vendor folder. It is not something that i generated manually)

What's your opinion?

**Comment by izzyblues** [comment]
@khatton-ft Forget about my comment above. I am going to use govendor properly in the project.

**Pull request merged**